// @chiragrupani/karma-chromium-edge-launcher/dist/headlessChannels/EdgeAnyHeadless.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseBrowser_1 = __importDefault(require("../BaseBrowser"));
const Utilities_1 = __importDefault(require("../Utilities"));
const EdgeAnyHeadlessBrowser = function (baseBrowserDecorator, args) {
    baseBrowserDecorator(this);
    var flags = args.flags || [];
    var userDataDir = args.edgeDataDir || this._tempDir;
    var browser = new BaseBrowser_1.default(flags, userDataDir);
    this._getOptions = browser._getHeadlessOptions;
};
EdgeAnyHeadlessBrowser.$inject = ['baseBrowserDecorator', 'args'];
EdgeAnyHeadlessBrowser.prototype = {
    name: 'Edge Any Headless',
    DEFAULT_CMD: {
        linux: Utilities_1.default.GetEdgeAnyLinux(),
        darwin: Utilities_1.default.GetEdgeAnyDarwin(),
        win32: Utilities_1.default.GetAnyEdgeWindows(),
    },
    ENV_CMD: 'EDGE_ANY_BIN',
};
exports.default = EdgeAnyHeadlessBrowser;
