// @chiragrupani/karma-chromium-edge-launcher/dist/headlessChannels/EdgeDevHeadless.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseBrowser_1 = __importDefault(require("../BaseBrowser"));
const Constants_1 = require("../Constants");
const Utilities_1 = __importDefault(require("../Utilities"));
const EdgeDevHeadlessBrowser = function (baseBrowserDecorator, args) {
    baseBrowserDecorator(this);
    var flags = args.flags || [];
    var userDataDir = args.edgeDataDir || this._tempDir;
    var browser = new BaseBrowser_1.default(flags, userDataDir);
    this._getOptions = browser._getHeadlessOptions;
};
EdgeDevHeadlessBrowser.$inject = ['baseBrowserDecorator', 'args'];
EdgeDevHeadlessBrowser.prototype = {
    name: 'Edge Dev Headless',
    DEFAULT_CMD: {
        linux: Utilities_1.default.GetLinuxBin(Constants_1.LinuxConstants.EdgeDev),
        darwin: Utilities_1.default.GetEdgeDarwin(`/Applications/${Constants_1.DarwinConstants.EdgeDev}.app/Contents/MacOS/${Constants_1.DarwinConstants.EdgeDev}`),
        win32: Utilities_1.default.GetEdgeExe(Constants_1.WindowsConstants.EdgeDev),
    },
    ENV_CMD: 'EDGE_DEV_BIN',
};
exports.default = EdgeDevHeadlessBrowser;
