// @chiragrupani/karma-chromium-edge-launcher/dist/channels/EdgeStable.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseBrowser_1 = __importDefault(require("../BaseBrowser"));
const Constants_1 = require("../Constants");
const Utilities_1 = __importDefault(require("../Utilities"));
const EdgeStableBrowser = function (baseBrowserDecorator, args) {
    baseBrowserDecorator(this);
    var flags = args.flags || [];
    var userDataDir = args.edgeDataDir || this._tempDir;
    var browser = new BaseBrowser_1.default(flags, userDataDir);
    this._getOptions = browser._getOptions;
};
EdgeStableBrowser.$inject = ['baseBrowserDecorator', 'args'];
EdgeStableBrowser.prototype = {
    name: 'Edge',
    DEFAULT_CMD: {
        linux: Utilities_1.default.GetLinuxBin(Constants_1.LinuxConstants.Edge),
        darwin: Utilities_1.default.GetEdgeDarwin(`/Applications/${Constants_1.DarwinConstants.Edge}.app/Contents/MacOS/${Constants_1.DarwinConstants.Edge}`),
        win32: Utilities_1.default.GetEdgeExe(Constants_1.WindowsConstants.Edge),
    },
    ENV_CMD: 'EDGE_BIN',
};
exports.default = EdgeStableBrowser;
